optuna/integration/lightgbm_tuner/optimize.py

```diff
@@ -533,10 +591,61 @@ def tune_params(self, target_param_names, n_trials, sampler):
         # Add tuning history.
         self.tuning_history += objective.report
 
-        if self.compare_validation_metrics(study.best_value, self.best_score):
-            self.best_score = study.best_value
+        best_trial = study.best_trial
+
+        # The values of complete trials are not None. The assertion is just for mypy.
+        assert best_trial.value is not None
+
+        if self.compare_validation_metrics(best_trial.value, self.best_score):
+            self.best_score = best_trial.value
             self.best_booster = objective.best_booster
 
-            updated_params = {p: study.best_trial.params[p] for p in target_param_names}
+            updated_params = {p: best_trial.params[p] for p in target_param_names}
             self.lgbm_params.update(updated_params)
             self.best_params.update(updated_params)
```

Comment:
Can we simplify this logic inside this `if`, making some of these `self.*` attribute `@property`s querying `self.study` (and `self.best_booster` for the `self.best_params`)? Not a strong opinion as as I still think we need this check of the booster is to be updated. But I otherwise believe it reduces code complexity.

Reply:
Thank you for your suggestion. I introduced `self.best_score` and `self.best_params` as properties, and simplified the logic.

The changes are as follows:
- I found that `objective.best_booster` is not `None` only if the trial outperformed the past best trial, and check it instead of calling `self.compare_validation_metrics`. I'd like to update this logic when LightGBMTuner supports the parallelization feature because the best booster can be updated by the other workers.
- The update of `self.lgbm_params` is omitted because it is updated in`self.tune_params` right before the training.
- `self.best_params` was renamed to `self._best_params`. The main logic is in the new `self.best_params` property and `self._best_params` is only used to keep backward compatibility.